### DISCUSSION TITLE
NiceGUI + AWS Lambda + Serverless Application

### DISCUSSION THREAD

**Discussion by johnlarkin1** [zauberzeug/nicegui#1449] [Q&A]
### Question

I am trying to set up a serverless application that sits on AWS Lambda that I can interact with from the browser through API Gateway. In other words, I'm imaging:

client (i.e. browser) -> api gateway -> aws lambda with deployed nice gui code that serves up response to browser. 

So what I'm imagining is basically I'll have some type of:

```py
from nicegui import ui

ui.run()

def lambda_handler() -> None:
    response = ui.render()  # get the HTML rendered from NiceGUI
    return {
        'statusCode': 200,
        'headers': {'Content-Type': 'text/html'},
        'body': response
    }
```

Is this type of setup with NiceGUI supported or not currently and I should use something more like Flask or Django?

**Comment by falkoschindler** [comment]
I don't have experience with AWS Lambda. But this response from ChatGPT sounds plausible:

> From the code snippet provided, it seems the intention is to use the NiceGUI library to render a user interface and return it through AWS Lambda. However, this might be problematic for several reasons:
>
> **Statefulness:** GUIs are typically stateful, meaning they maintain information about their current state (e.g., which buttons are clicked, what data is displayed). Lambda functions, on the other hand, are stateless; they don't maintain any internal state between invocations. Therefore, managing state between user interactions can be tricky.
>
> **Runtime Environment:** The AWS Lambda environment might not have all the required dependencies or system capabilities to run GUI-related code.
>
> **Response Size:** Lambda responses have size limitations. If the GUI code generates a large response, it might exceed the limit.

Since NiceGUI is based on FastAPI, this suggestion might be useful:

> FastAPI can be used with AWS Lambda and API Gateway, but you'll need a serverless framework like "Mangum" to adapt FastAPI applications for AWS Lambda & API Gateway.

So in summary:

> - AWS Lambda is best suited for stateless, event-driven applications.
> - For web applications, using frameworks like Flask, Django, or FastAPI (with the help of adapters like Mangum) is more common and might be easier.
> - The proposed NiceGUI approach might face challenges in the AWS Lambda environment, and I'd recommend exploring more conventional web frameworks.

**Reply by johnlarkin1** [reply]
Also curious - how does your ChatGPT know about NiceGUI? Browse-capability enabled? I thought the generally available one had its training data cutoff before the package was well established.

**Reply by falkoschindler** [reply]
This was my question to ChatGPT (no add-ons activated):

> Got this question on GitHub:
>
> [...]
>
> How does AWS lambda work? What is lambda_handler? Would it work with a simple FastAPI endpoint?

From this information it inferred that NiceGUI is a Web-UI framework, enough to give a very useful answer and explanations about integrating NiceGUI/FastAPI with AWS Lambda.

Sometimes I add some more information about NiceGUI (see https://github.com/zauberzeug/nicegui/wiki#chatgpt for example). But here I mainly wanted to learn about AWS Lambda. And since NiceGUI is based on FastAPI, it is often enough to ask about that instead if the question isn't about NiceGUI specifics.

**Reply by simplicity360** [reply]
Hi @falkoschindler , 
I think this is a very important aspect for NiceGUI because of its reliance on WebSockets and it's intended audience, right now, being upcoming small to medium sized deployments - whose focus would be on saving costs (lambda is very cheap for small businesses). There is a [WebSockets API integration with AWS Lambda](https://docs.aws.amazon.com/apigateway/latest/developerguide/websocket-api-chat-app.html) (official docs from AWS). [This reddit thread](https://www.reddit.com/r/aws/comments/124mfbt/how_expensive_is_it_using_lambdas_for_websocket/) discusses a lot more on advantages and limitations of this integration. 

[This API Gateway quotas document](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html#apigateway-execution-service-websocket-limits-table) lists limits on WebSockets frame size (32KB) and message payload size (128KB)

Perhaps it is possible to control the WebSocket message size of FastAPI or is this a topic that should be pursued in FastAPI forums?

**PS:** Stumbled upon this thread because I am in the process of moving away from the nicegui app that I created to go with tradition RestAPI + pure Frontend website OR Fullstack framework like Next.js. It's a shame because nicegui got my PoC kickstarted in a timescale not imaginable.

**Reply by evnchn** [reply]
Read https://aws.amazon.com/blogs/compute/choosing-between-aws-lambda-data-storage-options-in-web-apps/

It would be incompatible with the way NiceGUI works, which expects to have Python variables kept in memory, as well as multiple running `Client` instances, all of which would not be possible to be stored if we use AWS Lambda and a serverless implementation.

**Comment by tomjridge** [comment]
I too would like to implement a nicegui application running on lambda fronted by api gateway.

Using mangum with FastAPI, it is fairly straightforward for expose a FastAPI app over lambda and then to api gateway. I think the problem with nicegui is that it uses websockets for communication with the client (?not sure about this) and this may be difficult to get working with api gateway.

**Reply by rodja** [reply]
You are right. The websocket connection must be "sticky" (eg. connect back to the same server which delivered the initial request).

**Comment by baronne** [comment]
did you get anywhere with this?
